### PR TITLE
add id to metadata

### LINKED ISSUE #598: Add id to metadata
## Detailed Description
Add id to metadata, this could hold the gsp, pv system id number. This would be useful for the 'forecast' service

## Possible Implementation
Add it to [pydantic model](https://github.com/openclimatefix/nowcasting_dataset/blob/main/nowcasting_dataset/data_sources/metadata/metadata_model.py)

### PR DESCRIPTION
**Pull request by peterdudfield**
# Pull Request

## Description

Add id to metadata, this is optional so can be None. 

Fixes #598 #600 

## How Has This Been Tested?

normal unittests

- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by JackKelly — COMMENTED** [review]
This looks good.

One thought is:

In the near future (this month?), we plan to create batches across the entire geospatial extent of the satellite data (over the ocean; over Europe; over Africa; etc.). (These training examples won't have PV data, of course. Instead they'll be used to pre-train the model to predict the next few frames of satellite imagery).

And, separately, I also plan to create batches which are centered on individual PV systems (and where the model will know nothing about GSPs).

And we plan to produce start using PV data from mainland Europe and elsewhere.

In all three of these scenarios, there won't be a "GSP ID". As such, is the approach proposed in this PR is compatible with our plans to create batches which aren't associated with a GSP? When there isn't a GSP then I suppose we could just fill the `id` columns with NaNs or something like that, but that maybe feels like a bit of a hack?

Also, if we do stick with including the GSP ID, then I'd advocate for changing the name to `gsp_id` rather than just `id`, just to be really explicit?

**Comment by peterdudfield** [comment]
> This looks good.
> 
> One thought is:
> 
> In the near future (this month?), we plan to create batches across the entire geospatial extent of the satellite data (over the ocean; over Europe; over Africa; etc.). (These training examples won't have PV data, of course. Instead they'll be used to pre-train the model to predict the next few frames of satellite imagery).
> 
> And, separately, I also plan to create batches which are centered on individual PV systems (and where the model will know nothing about GSPs).
> 
> And we plan to produce start using PV data from mainland Europe and elsewhere.
> 
> In all three of these scenarios, there won't be a "GSP ID". As such, is the approach proposed in this PR is compatible with our plans to create batches which aren't associated with a GSP? When there isn't a GSP then I suppose we could just fill the `id` columns with NaNs or something like that, but that maybe feels like a bit of a hack?
> 
> Also, if we do stick with including the GSP ID, then I'd advocate for changing the name to `gsp_id` rather than just `id`, just to be really explicit?

Yea I agree it feels a bit of a hack. But yea that would be an option i.e set id to NaN or None. 

This is all work to get the forecast MVP working. Its because there is no GSP data to load, but I think we do need the GSP id for the MVP. One option would be to have some static GSP data, and do a hack when making batches/examples. 

Perhaps doing this: https://github.com/openclimatefix/nowcasting_dataset/issues/600, would help both tasks. Perhaps Ill give this a go, and itll actually make things tidier

**Comment by peterdudfield** [comment]
Would be something like this, and then `Location` can be used in other general places in the code

```
class Location(BaseModel):
    """ Location of the example"""
    
    t0_datetime_utc: pd.Timestamp = Field(
        ...,
        description="The t0 of one example ",
    )

    x_center_osgb: float = Field(
        ...,
        description="The x center of one example in OSGB coordinates",
    )

    y_center_osgb: float = Field(
        ...,
        description="The y center of one example in OSGB coordinates",
    )

    id: Optional[int] = Field(
        None,
        description="The id of the GSP or the PV system. "
        "This is optional so can be None",
    )


class Metadata(BaseModel):
    """Class to store metadata data"""

    batch_size: int = Field(
        ...,
        g=0,
        description="The size of this batch. If the batch size is 0, "
        "then this item stores one data item",
    )

    locations: List[Location]

 ```

**Comment by JackKelly** [comment]
Cool beans.

But, um, isn't the GSP ID (and PV system ID) already in each batch? I haven't checked the code recently but I definitely remember feeding both the GSP ID and PV system ID into my ML models back in December :slightly_smiling_face:  Or am I getting confused?

**Comment by peterdudfield** [comment]
> Cool beans.
> 
> But, um, isn't the GSP ID (and PV system ID) already in each batch? I haven't checked the code recently but I definitely remember feeding both the GSP ID and PV system ID into my ML models back in December 🙂 Or am I getting confused?

No your right! Just that means for the live models we would need the data sources: gsp in order for the system to work. Even though the model might not use gsp data. Im making good progress with #600 so hopefully itll all come together

File: nowcasting_dataset/consts.py [hunk 1]
```diff
@@ -133,6 +133,11 @@
 SPATIAL_AND_TEMPORAL_LOCATIONS_OF_EACH_EXAMPLE_FILENAME = (
     "spatial_and_temporal_locations_of_each_example.csv"
 )
-SPATIAL_AND_TEMPORAL_LOCATIONS_COLUMN_NAMES = ("t0_datetime_UTC", "x_center_OSGB", "y_center_OSGB")
+SPATIAL_AND_TEMPORAL_LOCATIONS_COLUMN_NAMES = (
```
Comment (peterdudfield) [inline nowcasting_dataset/consts.py:136]: remove this

File: nowcasting_dataset/data_sources/gsp/gsp_data_source.py [hunk 2]
```diff
@@ -146,6 +145,7 @@ def get_all_locations(
             1. list of datetimes
             2. list of x locations
             3. list of y locations
+            4. list of gsp ids
```
Comment (peterdudfield) [inline nowcasting_dataset/data_sources/gsp/gsp_data_source.py:148]: update return description

**Comment by peterdudfield** [comment]
Comments for chat with @JackKelly and @jacobbieker 
-change `Location` to `SpaceTimeLocation`
- add id_type, to show what type of id it is

File: nowcasting_dataset/data_sources/data_source.py [hunk 3]
```diff
@@ -378,9 +347,7 @@ def _get_time_slice(self, t0_datetime_utc: pd.Timestamp):
 
     def get_example(
         self,
-        t0_datetime_utc: pd.Timestamp,  #: Datetime of "now": The most recent obs.
-        x_center_osgb: Number,  #: Centre, in OSGB coordinates.
-        y_center_osgb: Number,  #: Centre, in OSGB coordinates.
+        location: SpaceTimeLocation,  #: Datetime of "now": The most recent obs.
```
Comment (jacobbieker) [inline nowcasting_dataset/data_sources/data_source.py:350]:
```suggestion
        location: SpaceTimeLocation,  #: Location object of the most recent observation
```

**Review by jacobbieker — APPROVED** [review]
Looks good to me! Just one small comment. Good job!

**Comment by peterdudfield** [comment]
@JackKelly im gona merge this, to push on the MVP, but please free feel to add comments, and ill do those changes

**Pull request merged**